### PR TITLE
[UC10#98] Implemented AcceptProposal RPC to allow users to accept a proposal and exchange physical cards

### LINKED ISSUE #98: Creare RPC per accettare la proposta di scambio, aggiornare i proprietari delle carte scambiate verificando che esistano ancora [UC10]
_**Dod**_
- [x]  La pull request ha un titolo e descrizione adeguata
- [x]  La pull request è stata collegata con la issue del task
- [x]  I requisiti funzionali del task collegato sono soddisfatti
- [x]  Il codice si compila senza errori
- [x]  Gli Unit Tests sono stati scritti e passati 
- [x]  Gli Unit Tests esistenti sono stati passati
- [x]  Il Code Review è stato fatto
- [ ]  La documentazione è stata aggiornata
- [x]  La pull request è stata accettata e fatto merge sul branch development.

 **Effort effettivo:**

### PR DESCRIPTION
**Pull request by alessioarcara**
This PR implements #98 and adds new RPC named `AcceptProposal` to allow users to accept a proposal and exchange physical cards. 
Also Refactored write operations in mapdb database to be transactional using WAL for crash protection.
The new method `writeOperation` now takes a callback as an argument and performs a commit if the callback is successful or a rollback if it fails. This approach reduces the complexity of the RPCs and helps keep them isolated, making them easier to test with unit tests.
To facilitate testing, a new class named FakeDB has been introduced. The FakeDB class implements the MapDB interface and simulates the database with fake operations, allowing for the testing of write operation callbacks as well.

### CODE REVIEW

File: src/main/java/com/aadm/cardexchange/client/presenters/ExchangeActivity.java [hunk 1]
```diff
@@ -47,6 +48,8 @@ public void onFailure(Throwable caught) {
                     view.showAlert(((AuthException) caught).getErrorMessage());
                 } else if (caught instanceof InputException) {
                     view.showAlert(((InputException) caught).getErrorMessage());
+                } else if (caught instanceof ProposalNotFoundException) {
+                    view.showAlert(((ProposalNotFoundException) caught).getErrorMessage());
```
Comment (Matte0Sacc0) [inline src/main/java/com/aadm/cardexchange/client/presenters/ExchangeActivity.java:52]: Nice to have a specific Exception for the missing proposal

File: src/main/java/com/aadm/cardexchange/server/mapdb/MapDBImpl.java [hunk 2]
```diff
@@ -39,4 +37,29 @@ public <K, V> Map<K, V> getPersistentMap(ServletContext ctx, String mapName, Ser
...
+            db.rollback();
+            return null;
+        }
+    }
+
+    @Override
+    public boolean writeOperation(ServletContext ctx, Runnable runnable) {
+        DB db = getDB(ctx, DB_FILE_TYPE);
+        try {
+            runnable.run();
```
Comment (Matte0Sacc0) [inline src/main/java/com/aadm/cardexchange/server/mapdb/MapDBImpl.java:58]: There has to be a commit to db also after this write operation

File: src/main/java/com/aadm/cardexchange/shared/exceptions/ExistingProposalException.java [hunk 3]
```diff
@@ -1,13 +1,13 @@
 package com.aadm.cardexchange.shared.exceptions;
 
 
-public class ExistingProposal extends BaseException {
+public class ExistingProposalException extends BaseException {
```
Comment (Matte0Sacc0) [inline src/main/java/com/aadm/cardexchange/shared/exceptions/ExistingProposalException.java:4]: I'd prefer to have this exception named "AlreadyExistingProposalException" as the way it is now appears to be a bit unclear. 

This won't affect the review tho

**Review by Matte0Sacc0 — COMMENTED** [review]
Really great job, i only request to add the commit after writeOperation in MapDBImpl

**Review by Matte0Sacc0 — APPROVED** [review]
Weel Done!!!

**Review by dstops — APPROVED** [review]

**Pull request merged**